### PR TITLE
[fsdp, megatron, vllm, trainer, algo, cfg] feat: Nitrobrew on-policy distillation (teacher hidden states communication + constant-memory fused KL)

### PR DESCRIPTION
**Pull request by tea-more**
### What does this PR do?

Adds **Nitrobrew**, a new on-policy distillation loss mode that estimates forward / reverse KL against the full teacher distribution **without ever materializing the `[N, V]` student-vs-teacher logit tensor** and **without communicating top-k logits**. Read our blogpost detailing the approach: [Nitrobrew: Fast, Lossless Distillation for Free](https://blog.tilderesearch.com/blog/nitrobrew).

Builds on the on-policy distillation infrastructure from #5041, #5723, #5745, #5997, #6051.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Amerged+distillation
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

**CPU unit tests** (`tests/experimental/teacher_loop/`)
- `test_nitrobrew_kernel_on_cpu.py`: chunked forward / reverse KL match a naive reference; autograd backward matches `torch.autograd.grad` on the reference.
- `test_nitrobrew_svd_on_cpu.py`: SVD round-trip and rank truncation, plus single-file / sharded / tied-embedding safetensors.

**GPU integration test**
- `tests/experimental/teacher_loop/test_nitrobrew_async_llm.py`: launches a `NitrobrewTeacherWorker` on Ray with `Qwen/Qwen3-0.6B`, runs end-to-end SVD → `AsyncLLM.encode` → on-actor projection, asserts shape `[S, d_comp]` and `bfloat16` dtype. Wired into `.github/workflows/vllm.yml`.

**End-to-end training experiment** — student `Qwen3-0.6B-Base`, teacher `Qwen3-4B`, dataset DAPO-Math-17k, 100 steps, 4 nodes × 8 GPUs each. Three runs at matched compute:

| Run | `loss_mode` | Communication budget per token |
|---|---|---|
| Nitrobrew (lossless, `d_comp = D`) | `nitrobrew` | 2560 floats |
| Top-k matched (`k = D`) | `forward_kl_topk` | 2 × 2560 floats (index + logprob) |
| Top-k practical (`k = 512`) | `forward_kl_topk` | 2 × 512 floats |

End-to-end step time:

> _[STEP TIME PLOT HERE — `nitrobrew-verl/logs/nitrobrew_e2e/step_time.png`]_

Headline: **Nitrobrew at full `d_model` is [SPEEDUP HERE]× faster end-to-end than the matched top-k baseline (k = 2560), and [SPEEDUP HERE]× faster than the practical top-k baseline (k = 512), while remaining fully lossless.**

Per-component breakdown (rollout / teacher inference / actor update):

> _[PER-COMPONENT BREAKDOWN PLOT HERE — `nitrobrew-verl/logs/nitrobrew_e2e/breakdown.png`]_

Loss / training curves (sanity check that the student is learning):

> _[LOSS CURVE PLOT HERE — `nitrobrew-verl/logs/nitrobrew_e2e/loss.png`]_

Model behaviour after 100 steps converges to the same regime as the topk baselines on DAPO-Math validation:

> _[VAL CURVE / TABLE HERE]_

For comprehensive cross-scale profiling and convergence results (Qwen3 0.6B/8B, 1.7B/32B, 4B/32B; up to 14× speedup on the largest pair), see [the blogpost](https://blog.tilderesearch.com/blog/nitrobrew).

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

New `loss_mode` plus per-teacher PCA rank — fully backward-compatible (no existing config field renamed):

```yaml
distillation:
  enabled: true
  distillation_loss:
    loss_mode: nitrobrew                  # NEW: option for nitrobrew loss mode
  teacher_models:
    teacher_model:
      model_path: Qwen/Qwen3-32B
      nitrobrew_d_comp: 256               # NEW: compression rank for this teacher
      inference:
        name: vllm
        tensor_model_parallel_size: 4
        gpu_memory_utilization: 0.85
        max_num_seqs: 128
```

### Design & Code Changes
**New files**   
- `verl/trainer/distillation/fsdp/nitrobrew_loss.py`: chunked KL + reverse-KL kernels; FSDP-side reduction.
- `verl/trainer/distillation/megatron/nitrobrew_loss.py`: Megatron tensor-parallel variant (vocab-sharded W_up).
- `verl/experimental/teacher_loop/nitrobrew_teacher.py`: `NitrobrewTeacherWorker` (Ray actor wrapping `AsyncLLM` in pooling+embed+ALL mode), `NitrobrewAsyncTeacherManager` (round-robin client with `asyncio.Lock`), `NitrobrewTeacherModelManager` (driver-side SVD + actor pool).
- `examples/on_policy_distillation_trainer/run_qwen3_nitrobrew_dapo.sh`: end-to-end OPD example covering all three comparison modes.
- `tests/experimental/teacher_loop/test_nitrobrew_kernel_on_cpu.py`
- `tests/experimental/teacher_loop/test_nitrobrew_svd_on_cpu.py`
- `tests/experimental/teacher_loop/test_nitrobrew_async_llm.py`  

**Modified files**  
- `verl/trainer/distillation/losses.py`: register nitrobrew loss mode + settings (`use_hidden_states=True`).
- `verl/workers/config/distillation.py`: add `nitrobrew_d_comp: int | None` to `DistillationTeacherModelConfig`; require it when `use_hidden_states`.
- `verl/workers/engine_workers.py`: `set_teacher_unembed(W)` registers W_up on actor ranks (vocab-shards on Megatron); uses `verl.utils.device.get_device_id()`.
- `verl/workers/utils/padding.py`: pad/unpad helpers for `[S, d_comp]` hidden-state batches.
- `verl/experimental/agent_loop/agent_loop.py`: accept `nitrobrew_worker_handles`, dispatch to `NitrobrewAsyncTeacherManager` when use_hidden_states.
- `verl/trainer/ppo/ray_trainer.py`: instantiate `NitrobrewTeacherModelManager` and call `actor_rollout_wg.set_teacher_unembed(w_up)` on the hidden-states branch.
- `.github/workflows/vllm.yml`: trigger paths and pytest step for the new GPU integration test.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

***AI Disclosure:** LLM coding assistants were used to scaffold the evaluation and testing harnesses for this PR.All AI-generated content was reviewed, edited, and approved by the human authors (@tea-more, @dhruvbpai) before submission.*

### CODE REVIEW

File: verl/experimental/agent_loop/agent_loop.py [hunk 1]
```diff
@@ -898,21 +928,29 @@ async def _compute_teacher_logprobs(
         validate: bool,
         sample_kwargs: Optional[dict[str, Any]] = None,
     ) -> None:
-        """Compute teacher logprobs for single sample."""
+        """Compute teacher signal for single sample (top-k logprobs or hidden states)."""
```
Comment (JacobHelwig) [inline verl/experimental/agent_loop/agent_loop.py]: Nit: we can return top-k logprobs or sampled logprob (for KL estimator such as k1). 
```suggestion
        """Compute teacher signal for single sample (logprobs or hidden states)."""
```

**Comment by wuxibin89** [comment]
Thanks for your contribution! A few comments:
1. Unify `NitrobrewTeacherModelManager` with existing [TeacherModelManager](https://github.com/verl-project/verl/blob/main/verl/experimental/teacher_loop/teacher_model.py#L36), and reuse [RolloutReplica](https://github.com/verl-project/verl/blob/main/verl/workers/rollout/replica.py#L70) for vllm/sglang/trtllm inference server.
3. Support multiple teachers(MOPD).
4. Support multi-modal VLM model.

**Comment by tea-more** [comment]
@wuxibin89 
Thanks for the reply.

I wanted to clarify what you mean by "unify with TeacherModelManager and reuse RolloutReplica". 
Do you mean that we shoul  (A) drop NitrobrewTeacherModelManager entirely; make it another flavor of the existing TeacherModelManager. Or (B) leep NitrobrewTeacherWorker as a standalone actor but make NitrobrewTeacherModelManager inherit from TeacherModelManager and reuse split_resource_pool + placement.

The reason it is separate right now is because the Nitrobrew teacher needs per-token hidden states, an on-actor projection step, and no weight-sync stack since the teacher is frozen.

**Comment by JacobHelwig** [comment]
Hi @tea-more. Full vocab distillation is a nice feature to add. 

Instead of adding new manager/replica classes, I am wondering if it is possible to cleanly make use/extend current classes to support full vocab distillation. We currently have these classes that have the same high-level functionality:

| Nitrobrew                      | verl                           |
|--------------------------------|--------------------------------|
| `NitrobrewTeacherModelManager` | `MultiTeacherModelManager`     |
| `NitrobrewAsyncTeacherManager` | `AsyncTeacherLLMServerManager` |
| `NitrobrewTeacherWorker`       | `RolloutReplica`               |

I am wondering if adding support for computing teacher hiddens in the `RolloutReplica` would allow for clean use of the existing managers in place of the new managers. This would allow us to drop all 3 of the Nitrobrew classes and add their functionality into the existing verl classes. 

> The reason it is separate right now is because the Nitrobrew teacher needs per-token hidden states, an on-actor projection step

I see 2 main differences in the Nitrobrew managers compared to verl:

1. Computation of compressed teacher head in `NitrobrewTeacherModelManager`
2. `NitrobrewAsyncTeacherManager.compute_teacher_hidden_states_single` $\approx$ `AsyncTeacherLLMServerManager.compute_teacher_logprobs_single` (except for method name and load balancing logic)

For 1, I think that it doesn't really belong in the manager, and moreso in the replica, both in terms of the level of abstraction as well as in terms of efficiency (possibly). I understand the reason for doing this is probably to avoid re-computing the teacher head SVD for all replicas. However, I think that this prevents the SVD from running on GPU, whereas I think the replica would be able to compute SVD using its allocated resources. I understand the SVD is a one-time cost at the beginning of training, but it runs once per teacher, which can cost several minutes on CPU depending on size and number of teachers:

image
 

For 2, we can re-name this method and other similar ones more abstractly to something like `acquire_teacher_knowledge_single` or `compute_teacher_signal_single`.

> no weight-sync stack since the teacher is frozen

Nice catch, this seems like an issue more general to the verl distillation implementation than Nitrobrew. Let's save it for another PR.

**Comment by dhruvbpai** [comment]
Hello @JacobHelwig 

I discussed with @tea-more and supporting naive SVD teacher head compression makes little sense at this point. We experimented with spectral compression strategies in our tech report and found the approach was quite poor - underperforming naive topk significantly. See `naive SVD` in figure below.

image

While we were able to develop more sophisticated variants that allowed for spectral compression, these require significantly more upfront computation time outside the trainer (e.g. sampling activations on held-out data and computing their covariance statistics). As such, spectral compression is likely out of scope for this PR.

@tea-more will submit changes to remove it accordingly, and if there is interest/the method is further developed, it can be added in a later PR.
